### ISSUE TITLE
ERROR:  query plan with multiple segworker groups is not supported when create extension postgis

### ISSUE THREAD

**Issue by taojin0505** [greenplum-db/gpdb#11304]
### Greenplum version or build
 PostgreSQL 9.4.24 (Greenplum Database 6.13.0 build commit:4f1adf8e247a9685c19ea02bcaddfdc200937ecd Open Source) on x86_64-unknown-linux-gnu, compiled by 
gcc (GCC) 6.4.0, 64-bit compiled on Dec 18 2020 22:31:16
### OS version and uname -a
LSB Version:    :core-4.1-amd64:core-4.1-noarch
Distributor ID: CentOS
Description:    CentOS Linux release 7.9.2009 (Core)
Release:        7.9.2009
Codename:       Core
Linux gyd-tc-app003 3.10.0-1160.6.1.el7.x86_64 #1 SMP Tue Nov 17 13:59:11 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
### Installation information ( pg_config )
BINDIR = /usr/local/greenplum-db-6.13.0/bin
DOCDIR = /usr/local/greenplum-db-6.13.0/share/doc/postgresql
HTMLDIR = /usr/local/greenplum-db-6.13.0/share/doc/postgresql
INCLUDEDIR = /usr/local/greenplum-db-6.13.0/include
PKGINCLUDEDIR = /usr/local/greenplum-db-6.13.0/include/postgresql
INCLUDEDIR-SERVER = /usr/local/greenplum-db-6.13.0/include/postgresql/server
LIBDIR = /usr/local/greenplum-db-6.13.0/lib
PKGLIBDIR = /usr/local/greenplum-db-6.13.0/lib/postgresql
LOCALEDIR = /usr/local/greenplum-db-6.13.0/share/locale
MANDIR = /usr/local/greenplum-db-6.13.0/man
SHAREDIR = /usr/local/greenplum-db-6.13.0/share/postgresql
SYSCONFDIR = /usr/local/greenplum-db-6.13.0/etc/postgresql
PGXS = /usr/local/greenplum-db-6.13.0/lib/postgresql/pgxs/src/makefiles/pgxs.mk
CONFIGURE = '--disable-debug-extensions' '--disable-tap-tests' '--enable-orca' '--with-zstd' '--with-gssapi' '--with-libxml' '--with-perl' '--with-python' '--with-openssl' '--with-pam' '--with-ldap' '--with-extra-version= Open Source' '--prefix=/usr/local/greenplum-db-6.13.0' '--mandir=/usr/local/greenplum-db-6.13.0/man' 'CC=gcc' 'CFLAGS=-O3 -fargument-noalias-global -fno-omit-frame-pointer -g'
CC = gcc
CPPFLAGS = -D_GNU_SOURCE -I/usr/include/libxml2 -I/usr/local/greenplum-db-6.13.0/include
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -O3 -fargument-noalias-global -fno-omit-frame-pointer -g -std=gnu99 -Werror=uninitialized -Werror=implicit-function-declaration -I/usr/local/greenplum-db-6.13.0/include
CFLAGS_SL = -fPIC
LDFLAGS = -Wl,--as-needed -Wl,-rpath,'/usr/local/greenplum-db-6.13.0/lib',--enable-new-dtags -L/usr/local/greenplum-db-6.13.0/lib
LDFLAGS_EX = 
LDFLAGS_SL = 
LIBS = -lpgcommon -lpgport -lxerces-c -lxml2 -lpam -lrt -lgssapi_krb5 -lzstd -lrt -lcrypt -ldl -lm -L/usr/local/greenplum-db-6.13.0/lib
VERSION = PostgreSQL 9.4.24
### Step to reproduce the behavior
1. compile and install postgis from (geospatial)[https://github.com/greenplum-db/geospatial],master branch
2. psql
3. postgres=# set optimizer=off;
4. postgres=# CREATE EXTENSION postgis;
ERROR:  query plan with multiple segworker groups is not supported
HINT:  likely caused by a function that reads or modifies data in a distributed table

**Comment by kainwen** [comment]
Can you upload more logs on master segment?

**Comment by taojin0505** [comment]
@kainwen 
thanks for your replying,this is the log
```
2020-12-23 17:33:37.904683 CST,"gpadmin","postgres",p22054,th2123831424,"[local]",,2020-12-23 17:33:33 CST,0,con14,cmd1,seg-1,,,,sx1,"LOG","00000","statement: set log_min_messages='debug1';",,,,,,"set log_min_messages='debug1';",0,,"postgres.c",1639,
2020-12-23 17:33:40.828652 CST,"gpadmin","postgres",p22054,th2123831424,"[local]",,2020-12-23 17:33:33 CST,0,con14,cmd1,seg-1,,,,,"DEBUG1","00000","Message type Q received by from libpq, len = 26",,,,,,,0,,"postgres.c",5226,
2020-12-23 17:33:40.828711 CST,"gpadmin","postgres",p22054,th2123831424,"[local]",,2020-12-23 17:33:33 CST,0,con14,cmd2,seg-1,,,,sx1,"LOG","00000","statement: CREATE EXTENSION postgis;",,,,,,"CREATE EXTENSION postgis;",0,,"postgres.c",1639,
2020-12-23 17:33:41.997562 CST,"gpadmin","postgres",p22054,th2123831424,"[local]",,2020-12-23 17:33:33 CST,744,con14,cmd5,seg-1,,dx18,x744,sx1,"ERROR","0A000","query plan with multiple segworker groups is not supported",,"likely caused by a function that reads or modifies data in a distributed table",,,,"CREATE EXTENSION postgis;",0,,"cdbdisp.c",288,
2020-12-23 17:33:41.997612 CST,"gpadmin","postgres",p22054,th2123831424,"[local]",,2020-12-23 17:33:33 CST,744,con14,cmd5,seg-1,,dx18,x744,sx1,"LOG","00000","An exception was encountered during the execution of statement: CREATE EXTENSION postgis;",,,,,,,0,,,,
2020-12-23 17:33:44.643827 CST,"gpadmin","postgres",p22054,th2122491648,"[local]",,2020-12-23 17:33:33 CST,0,con14,cmd5,seg-1,,,,,"LOG","00000","udp-ic: rx-thread shutting down",,,,,,,0,,,,
2020-12-23 17:33:44.807832 CST,"gpadmin","postgres",p22054,th2123831424,"[local]",,2020-12-23 17:33:33 CST,0,con14,cmd5,seg-1,,,,,"DEBUG1","00000","SessionState_Release: pinCount: 0, activeProcessCount: 0",,,,,,,0,,"session_state.c",153,"Stack trace:
1    0xbf0dac postgres errstart (elog.c:557)
2    0xc6c02d postgres SessionState_Shutdown (session_state.c:152)
3    0xc09d49 postgres <symbol not found> (postinit.c:1366)
4    0xa5f5c5 postgres shmem_exit (ipc.c:256)
5    0xa5f6f9 postgres proc_exit_prepare (ipc.c:216)
6    0xa5f7b1 postgres proc_exit (ipc.c:143)
7    0xa8a399 postgres PostgresMain (postgres.c:5664)
8    0x6b2195 postgres <symbol not found> (postmaster.c:4472)
9    0xa10fd2 postgres PostmasterMain (postmaster.c:1523)
10   0x6b5f21 postgres main (main.c:205)
11   0x7ffa7b54c555 libc.so.6 __libc_start_main + 0xf5
12   0x6c1c7c postgres <symbol not found> + 0x6c1c7c
"
2020-12-23 17:33:44.807914 CST,"gpadmin","postgres",p22054,th2123831424,"[local]",,2020-12-23 17:33:33 CST,0,con14,cmd5,seg-1,,,,,"DEBUG1","00000","Cleaning up Greenplum components...",,,,,,,0,,"cdbutil.c",1071,

```

**Comment by kainwen** [comment]
1. Please retry the create statement that throws the error you mentioned, and paste the log in segment (any segment should be OK, not master's log) here
2. Please verify the the two dirs' content is or is not the same, and paste answer here:
    * `$GPHOME/lib/postgresql` on master host machine
    * `$GPHOME/lib/postgresql` on segment host machine

With the above info, the issue should be clarified.

@taojin0505

**Comment by taojin0505** [comment]
@kainwen-zz 
Thanks for you comments,
1.  when I create extension,there's no log on segments
2.  file diff between master and segment:
```
# diff /usr/local/greenplum-db/lib/postgresql/ postgresql/
Common subdirectories: /usr/local/greenplum-db/lib/postgresql/pgxs and postgresql/pgxs
Binary files /usr/local/greenplum-db/lib/postgresql/postgis-2.5.so and postgresql/postgis-2.5.so differ
Binary files /usr/local/greenplum-db/lib/postgresql/rtpostgis-2.5.so and postgresql/rtpostgis-2.5.so differ
```
files on master 
```
# ls -l /usr/local/greenplum-db/lib/postgresql/postgis-2.5.so
-rwxr-xr-x 1 gpadmin gpadmin 4134616 Dec 23 14:58 /usr/local/greenplum-db/lib/postgresql/postgis-2.5.so
```
on segment:
```
# ls -l /usr/local/greenplum-db/lib/postgresql/postgis-2.5.so
-rwxr-xr-x 1 gpadmin gpadmin 4134616 Dec 23 14:52 /usr/local/greenplum-db/lib/postgresql/postgis-2.5.so
```

the master and segment  hosts both were new could instance, had the same os version.

**Comment by taojin0505** [comment]
@kainwen-zz  
is this [execute the below sqls,not "create extension postgis](https://github.com/greenplum-db/gpdb/issues/11390#issuecomment-764397851) a manual step to install postgis?

**Comment by taojin0505** [comment]
@kainwen-zz Many thanks.the problem fixed.When i install  rtpostgis.sql, it show error could not load "libgdal.so.20", and then dould there was a config error in /etc/ld.so.conf.

While when I run create extension,this error message not appear in log.
Thanks again!

**Comment by kainwen** [comment]
> @kainwen-zz Many thanks.the problem fixed.When i install rtpostgis.sql, it show error could not load "libgdal.so.20", and then dould there was a config error in /etc/ld.so.conf.
> 
> While when I run create extension,this error message not appear in log.
> Thanks again!

Greenplum does not report an understandable errmsg, we should fix this later.

**Comment by dusx1981** [comment]
I try to fix this issue in [#11683](https://github.com/greenplum-db/gpdb/pull/11683)

1:in cdbdisp_dispatchCommandInternal, if QE return err, the
cdbdisp_destroyDispatcherState which must be called is missing.

2:QE was set to the wrong state due to the execution error.

If we call cdbdisp_destroyDispatcherState directly
during error handling in cdbdisp_dispatchCommandInternal,
the original code called CdbDispatchUtilityStatement to
Reset QE CurrentExtensionObject, and QE is likely in abort
state, so the error return to client will be the QE error.

so we call DisconnectAndDestroyAllGangs to reset QE's state.
and PostgresMain will eventually call to cdbdisp_destroyDispatcherState
by calling AbortCurrentTransaction in the top of try catch stack.

**Comment by dusx1981** [comment]
I try to fix this in [#11683](https://github.com/greenplum-db/gpdb/pull/11683)

1:in cdbdisp_dispatchCommandInternal, if QE return err, the
cdbdisp_destroyDispatcherState which must be called is missing.

2:QE was set to the wrong state due to the execution error.

If we call cdbdisp_destroyDispatcherState directly
during error handling in cdbdisp_dispatchCommandInternal,
the original code called CdbDispatchUtilityStatement to
Reset QE CurrentExtensionObject, and QE is likely in abort
state, so the error return to client will be the QE error.

so we call DisconnectAndDestroyAllGangs to reset QE's state.
and PostgresMain will eventually call to cdbdisp_destroyDispatcherState
by calling AbortCurrentTransaction in the top of try catch stack.

**Comment by kainwen** [comment]
Fixed by https://github.com/greenplum-db/gpdb/commit/58628039c5369a520fc65adb87f0d29fd1203fd7 in master
and in 6x https://github.com/greenplum-db/gpdb/commit/4d185c05c8406cd38659a3aee952150c4cecb2b2
